testutils/infra/device.py

```diff
@@ -302,47 +301,68 @@ def _ssh_prep_args_impl(tool):
     return (cmd, host, port)
 
 
-def _put(file, local_path=".", remote_path="."):
-    (scp, host, port) = _scp_prep_args()
+def _put(device, file, local_path=".", remote_path="."):
+    (scp, host, port) = _scp_prep_args(device)
 
-    fabric.api.local(
+    subprocess.check_call(
         "%s %s %s/%s %s@%s:%s"
-        % (scp, port, local_path, file, fabric.api.env.user, host, remote_path)
+        % (scp, port, local_path, file, device.user, host, remote_path),
+        shell=True,
     )
 
 
-def _run(cmd, **kw):
+def _run(conn, cmd, **kw):
     if kw.get("wait") is not None:
         wait = kw["wait"]
         del kw["wait"]
     else:
         wait = 60 * 60
 
-    output = ""
+    result = None
     start_time = time.time()
     sleeptime = 1
-    # Use shorter timeout to get a faster cycle. Not recommended though, since
-    # in a heavily loaded environment, QEMU might be quite slow to use the
-    # connection.
-    with fabric.api.settings(timeout=60, abort_exception=Exception):
-        while True:
-            try:
-                output = fabric.api.run(cmd, **kw)
-                break
-            except Exception:
-                if time.time() >= start_time + wait:
-                    raise Exception("Could not successfully run: %s" % cmd)
-                time.sleep(sleeptime)
-                # Back off exponentially to save SSH handshakes in QEMU, which
-                # are quite expensive.
-                sleeptime *= 2
-                continue
-            finally:
-                # Taken from disconnect_all() in Fabric.
-                from fabric.state import connections
-
-                if connections.get(fabric.api.env.host_string) is not None:
-                    connections[fabric.api.env.host_string].close()
-                    del connections[fabric.api.env.host_string]
-
-    return output
+    while True:
+        # Back off exponentially to save SSH handshakes in QEMU, which
+        # are quite expensive.
+        time.sleep(sleeptime)
+        sleeptime *= 2
+
+        try:
+            result = conn.run(cmd, **kw)
+            break
+        except NoValidConnectionsError as e:
+            logger.info("Could not connect to host %s: %s", conn.host, e)
+            time.sleep(30)
```

Comment:
Why do you need an explicit sleep here when you already have `time.sleep(sleeptime)`? This will probably introduce unnecessary waiting time.

Reply:
Good point. It was a merge of two similar functions that resulted in double wait. I'll fix.